### PR TITLE
The same person should never be identified twice in the same photo

### LINKED ISSUE #376: Multiple faces in the same photo are being assigned to the same person
**Describe the bug**
When you click on a face in the People view in Photos, The same photo is listed multiple times. For example I have photos from my daughter's choir performance with about 100 kids in the shot. When I click on a face in the People view in Photos many photos of this choir performance are shown, with the same photo shown many, perhaps dozens of times. If you click the select box on one of the photos, all the copies of that photo are selected. There is no way to know which face in each photo is assigned to this person. (no bounding box is displayed). If you select "Remove photo from person", then go back to Photos, back to People, and back to this person, the same photo will still be shown, just one fewer copies of it. There doesn't seem to be a way to select which copy will be removed.

**To Reproduce**
1. Go to Photos
2. Click on People
3. Click on a person's face
4. All the photos containing this person are listed. Some photos with large groups may be listed many times.

**Expected behavior**
Each photo containing this person's face should be listed once.
The same person should never be identified twice in the same photo.
The person's face should be highlighted in some way so that incorrectly identified faces can accurately be removed from the person.

**Recognize (please complete the following information):**
 - JS-only mode: [e.g. Yes/No] Not shown in app settings
 - Enabled modes: [e.g. object detection, face recognition]
Face Recognition: on
Object recognition: on
Landmark recognition: on
Audio tagging: off
Video tagging: off
WASM mode: off

**Server (please complete the following information):**
 - Nextcloud: 25.0.0 RC3
 - Recognize App V 3.0.1
 - OS: Ubuntu 22.04
 - RAM: 8GB
 - Processor Architecture Intel I7 64 bit

### PR DESCRIPTION
**Pull request by marcelklehr**
FaceClusterAnalyzer: Knock out less matching faces from a cluster if they are on the same file

fixes #376

### CODE REVIEW

File: lib/Service/FaceClusterAnalyzer.php [hunk 1]
```diff
@@ -100,10 +100,99 @@ public function calculateClusters(string $userId) {
...
+			}
+		}
+
+		$this->pruneClusters($userId);
+	}
+
+	/**
+	 * @throws \OCP\DB\Exception
+	 */
+	public function pruneClusters(string $userId) {
```
Comment (come-nc) [inline lib/Service/FaceClusterAnalyzer.php:112]:
```suggestion
	public function pruneClusters(string $userId): void {
```

File: lib/Service/FaceClusterAnalyzer.php [hunk 2]
```diff
@@ -100,10 +100,99 @@ public function calculateClusters(string $userId) {
...
+			 * @var $detections FaceDetection[]
+			 */
+			$detections = $this->faceDetections->findByClusterId($cluster->getId());
+
+			$filesWithDuplicateFaces = $this->findFilesWithDuplicateFaces($detections);
+			if (count($filesWithDuplicateFaces) === 0) {
+				continue;
+			}
+
+			$centroid = $this->calculateCentroidOfDetections($detections);
```
Comment (come-nc) [inline lib/Service/FaceClusterAnalyzer.php:134]: Can you explain what a centroid is and what is the selection logic below?
Are all $detections here of the same face?

How does the code avoids being confused in the case of multiple detection of several faces in the same image? A cluster is specific to a person and so we only work with detections of the same face in all this block, right?

File: lib/Service/FaceClusterAnalyzer.php [hunk 3]
```diff
@@ -100,10 +100,99 @@ public function calculateClusters(string $userId) {
...
+			}
+			$files[$detection->getFileId()][] = $detection;
+		}
+
+		/** @var array<int,FaceDetection[]> $filesWithDuplicateFaces */
+		$filesWithDuplicateFaces = array_filter($files, function ($detections) {
+			return count($detections) > 1;
+		});
+
+		return$filesWithDuplicateFaces;
```
Comment (come-nc) [inline lib/Service/FaceClusterAnalyzer.php:196]:
```suggestion
		return $filesWithDuplicateFaces;
```

**Review by come-nc — COMMENTED** [review]
Obama crowning himself meme

File: lib/Service/FaceClusterAnalyzer.php (hunk 2 above)
Comment (marcelklehr) [reply lib/Service/FaceClusterAnalyzer.php:134]: > A cluster is specific to a person and so we only work with detections of the same face in all this block, right?

Yes, then we calculate the arithmetic mean (the centroid) of all face vectors in this cluster. Then we group face detections by file id and for all file ids keep only the face detection that is closest to the centroid.

**Comment by marcelklehr** [comment]
I know that it's technically possible for the same face to appear in the same photo multiple times, but statistically it's a hard outlier and in the typical case should be regarded as a false positive, IMO. A different strategy would be to keep the duplicate face detections in the db but only return one of the detections in the DAV endpoint. I found this approach cleaner.

File: lib/Service/FaceClusterAnalyzer.php (hunk 2 above)
Comment (come-nc) [reply lib/Service/FaceClusterAnalyzer.php:134]: Ok so the centroid is computed on all detections and not only the ones on the files with duplicates, so it makes sense to take the one closest to the centroid.

**Review by come-nc — APPROVED** [review]
The image was a joke, not a complain :-)
Even if the face is actually multiple times in the file it’s useless to make the face appear several times in the list, and the linked tickets shows that it causes problems to keep duplicates.
It would make no sense to deterior usablility in normal cases just to support the edge case of fake photo or mirror house.

**Comment by marcelklehr** [comment]
Thank you for reviewing :hearts:

**Pull request merged**